### PR TITLE
Zero pad rounds fix

### LINKED ISSUE #1563: Insufficient zero-padding bug for functions returning byte arrays of size < 16
### Version Information

* vyper Version (output of `vyper --version`): 0.1.0b11+commit.c775cda (the latest master)
  * The buggy behavior happens for both recent releases, [v0.1.0-beta.10](https://github.com/ethereum/vyper/releases/tag/v0.1.0-beta.10) and [v0.1.0-beta.11](https://github.com/ethereum/vyper/releases/tag/v0.1.0-beta.11), as well.

#### Summary:

A function whose return type is `bytes[n]` where `n < 16`, returns a value that does *not* conform to [the ABI encoding](https://solidity.readthedocs.io/en/v0.5.10/abi-spec.html), having incorrect zero-padding.

#### Details:

For example, [the `get_deposit_count` function of the deposit contract](https://github.com/ethereum/eth2.0-specs/blob/v0.8.0/deposit_contract/contracts/validator_registration.v.py#L60), whose return type is `bytes[8]`, returns the following 96 bytes (in hexadecimal notation):
```
0x0000000000000000000000000000000000000000000000000000000000000020
  0000000000000000000000000000000000000000000000000000000000000008
  deadbeefdeadbeef000000000000000000000000000000000000000000000020
                                                                ^^  <-- problem!
```
where the first 32 bytes (the first line) denotes the header (the offset 32 = 0x20), the second 32 bytes (the second line) denotes the length of the byte array (8), and the "`deadbeefdeadbeef`" denotes the content of the byte array (`bytes[8]`). *Here, the problem is that the last byte is 32 (= 0x20) while it should be 0 (= 0x00) according to the ABI specification.* This means that any contract (written in either Solidity or Vyper) that calls to this type of functions expecting the correct zero-padding may misbehave.

What happens is that the compiled bytecode of `get_deposit_count` fails to sufficiently pad zeros when it prepares for the return value. In the following return statement,
```
return self.to_little_endian_64(self.deposit_count)
```
it first copies the returned value (of 8 bytes) to some specific region of the memory, and puts only 8 bytes of zero-padding after that, instead of 24 bytes of zeros, which results in including some garbage values in the last 16 bytes. Indeed, in this particular case, the last 16 bytes (`000000000000000000000000020`) came from the side-effect of the sub-function call to `self.to_little_endian_64()`.

To be more specific, in the following zero-padding LLL code of `get_deposit_count` (generated by [the `zero_pad` function of `vyper/parser/stmt.py`](https://github.com/ethereum/vyper/blob/c775cda6a34b62612b4160d9017ad701f4f52db8/vyper/parser/stmt.py#L780-L799)):
```
            /* Zero pad */
            [with,
              _ceil32_end,
              [ceil32, [mload, 640]],
              [repeat,
                736,
                [mload, 640],
    ??? -->     8,
                [seq,
                  [if, [gt, [mload, 736], _ceil32_end], break],
                  [mstore8, [add, 672, [mload, 736]], 0]]]],
            [mstore, 608, 32],
            [return, 608, [ceil32, [add, [mload, 640], 64]]],
            # Line 58
            stop]],
```
the third argument of the `repeat` loop is 8, where it should be at least 24.  In a quick examination of the Vyper compiler code, it seems that the number `8` comes from the `maxlen` of the type `bytes[8]`, but I couldn't have a chance to further investigate the root cause and a quick fix.

Indeed, the same problem happens in [the `to_little_endian_64` function](https://github.com/ethereum/eth2.0-specs/blob/v0.8.0/deposit_contract/contracts/validator_registration.v.py#L30) as well,  which has less effect because it is a private function, though.

_[This bug was privately reported on July 22, 2019, confirmed as not a security vulnerability, and made public here for the transparency.]_

### PR DESCRIPTION
**Pull request by jacqueswww**
### What I did
Fixes #1563
### Cute Animal Picture

🐀

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM python:3.6-slim
+FROM python:3.6-slim as Build
```
Comment (charles-cooper) [inline Dockerfile]: not sure this should be in this PR

Comment (jacqueswww) [reply Dockerfile]: oh right, slipped in - will fix

**Comment by charles-cooper** [comment]
Two things bother me about this, one is that there is repeated code (also in events.py which might also have the same bug in it - haven't investigated). The other thing, and I'm on the fence about this, is that use of mstore8. It will definitely be more efficient, and possibly easier to reason about, if we use word operations for zeroing. For the edge case we can use the following bitmasking operation

```python
# pseudocode
actual_len = mload (mload (bytes_ptr)) 
xtra = actual_len % 32
last_word_ptr = actual_len - xtra
zero_padded_last_word = mload(last_word_ptr) & (2**xtra - 1)
mstore(last_word_ptr, zero_padded_last_word)

# .. zero out the rest between last_word_ptr / 32 + 1 and maxlen.
```

@jacqueswww what do you think?

**Comment by jacqueswww** [comment]
> Two things bother me about this, one is that there is repeated code (also in events.py which might also have the same bug in it - haven't investigated). The other thing, and I'm on the fence about this, is that use of mstore8. It will definitely be more efficient, and possibly easier to reason about, if we use word operations for zeroing. For the edge case we can use the following bitmasking operation
> 
> ```python
> # pseudocode
> actual_len = mload (mload (bytes_ptr)) 
> xtra = actual_len % 32
> last_word_ptr = actual_len - xtra
> zero_padded_last_word = mload(last_word_ptr) & (2**xtra - 1)
> mstore(last_word_ptr, zero_padded_last_word)
> 
> # .. zero out the rest between last_word_ptr / 32 + 1 and maxlen.
> ```
> 
> @jacqueswww what do you think?

Yes I think that's good idea, although I would opt to make that change after this is merged. `events.py` seems to already have the necessary ceil32. https://github.com/ethereum/vyper/blob/master/vyper/parser/events.py#L111

**Comment by davesque** [comment]
@jacqueswww This will certainly zero out the correct number of bytes.  However, is there any way to be confident that the bytes that will now be zeroed aren't being used for something else?  Is it correct that this would depend on whether or not the allocator counted the required number of bytes for the type correctly?  If so, what sections of the compiler would I look at to make sure this is the case?

**Comment by jacqueswww** [comment]
I am not sure what you mean? The len bytes are allocated at assignment. And the zero bytes are not used because the abi uses 32 byte boundaries.

**Comment by davesque** [comment]
Actually, I think I found it here:
https://github.com/ethereum/vyper/blob/984aed51fd18f4b8bd1c8465c249035ac08a1ac9/vyper/types/types.py#L492

I'm curious why the required number of words is `ceil32(maxlen) // 32 + 2`.  I would expect it to be `ceil32(maxlen) + 1` where the `1` represents the length prefix word.

**Comment by davesque** [comment]
> Vyper uses 32 byte word base for every type

Ahh, right.

**Comment by jacqueswww** [comment]
I agree we have to actually separate out the sizes there into memory size & abi size, that one looks like abi size - which is obviously incorrect.

I think the restriction you are looking for get enforced here:
https://github.com/ethereum/vyper/blob/master/vyper/parser/memory_allocator.py#L25

**Comment by davesque** [comment]
Yeah, I can see from that link you provided that a particular value will only ever begin on a word boundary.  That's nice and that makes things easier to reason about.

However, whether or not the length prefix words are being properly calculated for by `get_size_of_type`, I don't see those prefix words being accounted for by any of the `zero_pad` routines.  Maybe that's being done somewhere else and it's not a problem, but would that cause any issues?

**Comment by jacqueswww** [comment]
It would also cause over allocation, except for extra gas cost, it would cause trouble. Also IIRC the EVM requires continuous memory allocation - that would've have thrown.

**Comment by davesque** [comment]
> Also IIRC the EVM requires continuous memory allocation - that would've have thrown.

Do you mean that the EVM would have returned an error status somehow for misallocated memory?  I'm not sure that's the case.  I think it basically just does `ceil32(addr)` where `addr` is the highest memory address referenced by an op that mutates memory and uses that as the basis for calculating the memory expansion cost.  In other words, allocation is done implicitly and on demand, and the EVM doesn't care if memory is referenced continuously or not, and it wouldn't throw for any reason related to that.

**Comment by jacqueswww** [comment]
Ah OK, either way I reckon those could be solved in other PRs if necessary. For this one it should just introduce bettwr tests and the basic fix. Until we get something betterbas Xharles suggested.

**Comment by davesque** [comment]
Actually, I found the commit that introduced that change: 35ae7c178d29fc414d467ce254e81672e0f94d6f

Not sure how that allows for having "slice support".

**Comment by charles-cooper** [comment]
> I'm curious why the required number of words is ceil32(maxlen) // 32 + 2. I would expect it to be ceil32(maxlen) // 32 + 1 where the 1 represents the length prefix word.

Yeah I thought that was weird too but it seems to be working. You could try changing it to `ceil32(maxlen)//32 + 1` and see if that breaks anything. Also yeah EVM doesn't require memory to be allocated continuously so this wouldn't have caused any errors if it overallocated, just marginal 3 extra gas for memory expansion.

**Comment by charles-cooper** [comment]
> Ah OK, either way I reckon those could be solved in other PRs if necessary. For this one it should just introduce better tests and the basic fix. Until we get something better as Charles suggested.

I would feel a lot more comfortable if we could factor out the repeated code now, otherwise it might never get refactored until we find another bug from having repeated code that evolves in slightly different ways.

**Comment by jacqueswww** [comment]
Will do.

**Comment by jacqueswww** [comment]
@charles-cooper refactored, I think this is ready to be merged now?

**Comment by charles-cooper** [comment]
Much cleaner - thanks @jacqueswww ! LGTM

File: vyper/parser/events.py [hunk 2]
```diff
@@ -108,24 +109,9 @@ def pack_args_by_32(holder, maxlen, arg, typ, context, placeholder,
...
-                            ['add', ['add', '_bytearray_loc', 32], ['mload', zero_pad_i]],
-                            0,
-                        ],
-                    ]],
-                ]],
-            ]
-        ])
+        holder.append(
+            zero_pad(dest_placeholder, maxlen, zero_pad_i=zero_pad_i)
+        )
```
Comment (daejunpark) [inline vyper/parser/events.py]: @jacqueswww I do not think this refactoring is semantics-preserving.

In order to be semantics-preserving, the following two conditions need to be satisfied, but I have a counter-example that the first condition is not always met.

1. `maxlen` is equal to `new_maxlen`.
2. The runtime value of `_ceil32_end` is equal to `new_maxlen`.

Comment (daejunpark) [reply vyper/parser/events.py]: @charles-cooper I forgot to mention you for this. What do you think about this?

Comment (charles-cooper) [reply vyper/parser/events.py]: I'm not sure. I thought `maxlen` will get `ceil32` applied to it in `zero_pad`.

**Pull request merged**